src/pool/srv_pool.c

```diff
@@ -2699,145 +2818,285 @@ pool_op_is_write(crt_opcode_t opc)
 	return is_write;
 }
 
+#if 0
+/* DEBUG */
+static int
+pool_op_iter_cb(daos_handle_t ih, d_iov_t *key_enc, d_iov_t *val, void *arg)
+{
+	struct ds_pool_svc_op_key  op_key;
+	struct ds_pool_svc_op_val *op_val = val->iov_buf;
+
+	ds_pool_svc_op_key_decode(key_enc, &op_key);
+
+	D_DEBUG(DB_MD, "key: time=" DF_X64 ", cli=" DF_UUID ", rc=%d\n",
+		op_key.ok_client_time, DP_UUID(op_key.ok_client_id), op_val->ov_rc);
+
+	return 0;
+}
+#endif
+
+static int
+pool_op_check_delete_oldest(struct rdb_tx *tx, struct pool_svc *svc, bool dup_op,
+			    uint32_t *svc_ops_num)
+{
+	int                       rc;
+	d_iov_t                   key1_enc;
+	struct ds_pool_svc_op_key k1;
+	uint64_t                  t1_sec;
+	uint64_t                  t2_sec;
+	uint64_t                  age_sec;
+	uint32_t                  num_deleted = 0;
+	bool                      need_delete = false;
+
+	if (svc->ps_ops_enabled == 0)
+		return 0;
+
+	d_iov_set(&key1_enc, NULL, 0);
+	rc = rdb_tx_fetch(tx, &svc->ps_ops, RDB_PROBE_FIRST, NULL /* key_in */, &key1_enc,
+			  NULL /* value */);
+	if (rc == -DER_NONEXIST)
+		return 0;
+	else if (rc != 0) {
+		DL_ERROR(rc, "failed to probe first ps_ops entry");
+		return rc;
+	}
+
+	rc = ds_pool_svc_op_key_decode(&key1_enc, &k1);
+	if (rc != 0) {
+		DL_ERROR(rc, "key decode failed");
+		return rc;
+	}
+
+	/* If number of RPCs is at the limit, or the oldest is more than 5 minutes old,
+	 * delete the oldest entry. TODO: evict many/all such entries (during periodic cleanup?).
+	 */
+	t1_sec  = d_hlc2sec(k1.ok_client_time);
+	t2_sec  = d_hlc2sec(d_hlc_get());
+	age_sec = t2_sec - t1_sec;
+
+	if ((*svc_ops_num >= svc->ps_ops_max) || (age_sec > svc->ps_ops_age))
+		need_delete = true;
```

Comment:
[style] codes might more clean:

if (*svc_ops < svc->ps_ops_max && age_sec <= svc->ps_ops_age)
               return;

rc = rdb_tx_delete()
if (rc != 0)
    ....

*svc_ops_num--;

we don't need @need_delete and @num_deleted